usim_pytest/test_types/test_locks.py

```diff
@@ -0,0 +1,116 @@
+import pytest
+
+from usim import time, Lock, Scope, instant
+
+from ..utility import via_usim
+
+
+class TestLock:
+    @via_usim
+    async def test_misuse(self):
+        with pytest.raises(AttributeError):
+            with Lock():
+                ...
+
+    @via_usim
+    async def test_acquire_single(self):
+        lock = Lock()
+        async with lock:
+            await (time + 5)
+        assert time == 5
+
+    @via_usim
+    async def test_acquire_reentry(self):
+        lock = Lock()
+        async with lock:
+            async with lock:
+                async with lock:
+                    await (time + 5)
+        async with lock, lock, lock:
+            await (time + 5)
+        assert time == 10
+
+    @via_usim
+    async def test_acquire_multiple(self):
+        lock_a, lock_b, lock_c = Lock(), Lock(), Lock()
+        async with lock_a:
+            async with lock_b:
+                async with lock_c:
+                    await (time + 5)
+        async with lock_a, lock_b, lock_c:
+            await (time + 5)
+        assert time == 10
+
+    @via_usim
+    async def test_contended(self):
+        lock = Lock()
+
+        async def mutext_sleep(delay):
+            async with lock:
+                await (time + delay)
+        async with Scope() as scope:
+            scope.do(mutext_sleep(5))
+            scope.do(mutext_sleep(5))
+            scope.do(mutext_sleep(10))
+        assert time == 20
+
+    @via_usim
+    async def test_available(self):
+        lock = Lock()
+
+        async def hold_lock():
+            async with lock:
+                await (time + 10)
+        assert lock.available
+        async with lock:
+            assert lock.available
+        async with Scope() as scope:
+            scope.do(hold_lock())
+            await (time + 5)
+            assert not lock.available
+        assert lock.available
+
+    @via_usim
+    async def test_release_exception(self):
+        lock = Lock()
+
+        with pytest.raises(KeyError):
+            async with lock:
+                raise KeyError
+        with pytest.raises(KeyError):
+            async with lock:
+                raise KeyError
```

Comment:
Why do you execute the check for KeyEorror two times?

Reply:
The second test is intended to check that the lock was released properly. It could also use an ``assert True`` in the body, though.

Would that make it more clear what is going on?

Reply:
The `assert True` comes behind. Just leave as is.